Fixed #12952 -- Made admin's model history change messages use form labels instead of field names.

CuriousLearner opened the pull request:
cc @carltongibson 


PS: Special Thanks to @andrewgodwin helping in resolving the issue around serializing the gettext versioned proxy objects during DjangoCon US sprints :)

claudep commented on django/contrib/admin/utils.py:
```diff
@@ -490,10 +492,17 @@ def construct_change_message(form, formsets, add):
     change_message = []
+    changed_fields_labels = [_(form.fields[f].label) for f in form.changed_data if f != 'password']
+    if 'password' in form.changed_data:
+        # password field is a special case. Form fields are `password1` and `password2`
```
I'm not sure special-casing the password field is a good idea, as you might as well have user code presenting such cases, and then the function would crash badly. I think a try/except clause catching the KeyError when f is not find in form.fields, defaulting to the raw field name in that case, would be a safer approach.

claudep commented on tests/admin_utils/test_logentry.py:
```diff
@@ -53,9 +53,9 @@ def test_logentry_change_message(self):
+        self.assertEqual(logentry.get_change_message(), 'Changed Title and History.')
         with translation.override('fr'):
-            self.assertEqual(logentry.get_change_message(), 'Modification de title et hist.')
+            self.assertEqual(logentry.get_change_message(), 'Modification de Title et History.')
```
It should give 'Modification de Title et Historique.'. I guess a gettext call is missing inside the `LogEntry.get_change_message`.

claudep commented on tests/admin_utils/test_logentry.py:
```diff
@@ -85,7 +85,7 @@ def test_logentry_change_message_localized_datetime_input(self):
             self.assertRedirects(response, reverse('admin:admin_utils_article_changelist'))
         logentry = LogEntry.objects.filter(content_type__model__iexact='article').latest('id')
-        self.assertEqual(logentry.get_change_message(), 'Changed title and hist.')
+        self.assertEqual(logentry.get_change_message(), 'Changed Title and Historique.')
```
`and Historique` does not look right. It should give the untranslated `History` here (we are outside of the override).

CuriousLearner replied:
Yeah, fixed this!

CuriousLearner commented:
Hi @claudep 

I've addressed your reviews. The failure of sqlite3 DB seems unrelated to the changes.

Can you please take another pass here?

claudep commented on tests/admin_views/tests.py:
```diff
@@ -2032,6 +2032,29 @@ def test_history_view(self):
 
                 self.client.get(reverse('admin:logout'))
 
+    def test_history_view_changed_fields_message(self):
```
I think the test is not in the right class (`AdminViewPermissionsTest` should be limited to test permission-related stuff). We may need a new test class.

CuriousLearner commented:
@claudep I see. For the first diff, what do you think should be a good way to handle that? Should we give out an error that such field does not even exist, or just suppress it silently, and just giving out the label name for fields that actually exist on the model?

claudep commented:
That was my initial remark on the `KeyError`. Instead of the 'password' special case, I would write something like (field_name instead of f):
```
            try:
                verbose = form.fields[field_name].label or field_name
            except KeyError:
                verbose = field_name
            change_field_labels.append(verbose)
```

CuriousLearner commented:
Hey @claudep 

Does this looks okay now?

claudep commented on django/contrib/admin/utils.py:
```diff
@@ -12,7 +12,9 @@
 from django.utils.text import capfirst
-from django.utils.translation import ngettext, override as translation_override
+from django.utils.translation import (
+    ngettext, override as translation_override, ugettext as _,
```
Use `gettext` instead of `ugettext`

CuriousLearner replied:
I read at various places that `ugettext` is often recommended than `gettext`. Is there specific reason we're converting it to `gettext` here?

(Although it won't be needed in next version of patch, but I just wanted to know the rationale behind this :) )

claudep replied:
It used to be that way (in Python 2 era). Now gettext is an alias to ugettext and the latter will be deprecated in the future.

CuriousLearner replied:
Oh, I see. Thank you for answering my query :)

claudep commented on django/contrib/admin/utils.py:
```diff
@@ -506,11 +518,12 @@ def construct_change_message(form, formsets, add):
                         }
                     })
                 for changed_object, changed_fields in formset.changed_objects:
+                    changed_fields_labels = [_(formset.forms[0].fields[f].label) for f in changed_fields]
```
Didn't you have a version of your patch where you called a method to get the labels? I guess we might have the same issues here than above (label=None or field name not in fields).

CuriousLearner replied:
Yeah, I'll fix this one as well.

claudep commented on django/contrib/admin/utils.py:
```diff
@@ -490,10 +492,20 @@ def construct_change_message(form, formsets, add):
+    with translation_override(None):
+        for field_name in changed_data:
+            try:
+                verbose_field_name = _(form.fields[field_name].label) or field_name
```
I may miss something, but why calling gettext if we are in a block with translations deactivated?

CuriousLearner replied:
Ah, it was mostly to have it later converted to an `str` to dumps as json. I think it would be better to just use `str` here.

claudep commented on tests/admin_views/admin.py:
```diff
@@ -881,8 +881,16 @@ class CityInlineAdmin(admin.TabularInline):
     view_on_site = False
 
 
+class StateAdminForm(forms.ModelForm):
```
I suggested a field without label in that form, don't you find it useful to test fields without labels? Or is it tested elsewhere?

timgraham replied:
If there's a purpose for `label_form_field` at this point, I can't easily spot it.

CuriousLearner replied:
Thanks for pointing this out Tim. I actually forgot to check this in the test and as soon as I added this, I discovered another issue.

We were doing: 
`verbose_field_name = (form.fields[field_name].label) or field_name`

which would make `field_name` unreachable. Since, even if `form.fields[field_name].label` is `None`, the stringify version of it would always evaluate to `True`.
I'm fixing this as well as the similar change in case formsets are there.

claudep commented on tests/admin_views/tests.py:
```diff
@@ -6069,3 +6069,45 @@ def test_implicitly_generated_pk(self):
+            username='super', password='secret', email='super@example.com'
+        )
+        # login POST dicts
+        cls.index_url = reverse('admin:index')
```
I don't think this and the lines below are useful. We can simply use `force_login` in the test as far as the login process itself is not significant in that test.

claudep commented on tests/admin_views/tests.py:
```diff
@@ -6069,3 +6069,45 @@ def test_implicitly_generated_pk(self):
+        article_change_url = reverse('admin:admin_views_state_change', args=(state.pk,))
+        self.client.login(**self.super_login)
+        self.client.post(article_change_url, change_dict)
+        response = self.client.get(reverse('admin:admin_views_state_history', args=(state.pk,)))
```
I wonder if testing the `LogEntry` items instead of calling the view would be a bit more precise testing. To be evaluated.

CuriousLearner replied:
Yeah, that actually makes more sense. I would rather test the `LogEntry` change message here for the changed fields.

CuriousLearner commented:
@claudep Can you please take another pass here?

Thank you for your patience.

carltongibson commented on django/contrib/admin/utils.py:
```diff
@@ -506,11 +516,19 @@ def construct_change_message(form, formsets, add):
+                        try:
+                            verbose_field_name = str(formset.forms[0].fields[field_name].label) or field_name
+                        except KeyError:
+                            verbose_field_name = field_name
```
This `except` block is never tested in the `formsets` case.

CuriousLearner replied:
Fixed :)

claudep commented on tests/admin_views/tests.py:
```diff
@@ -6069,3 +6069,39 @@ def test_implicitly_generated_pk(self):
+
+    def test_history_view_changed_fields_message(self):
+        """
+        Test the history_view change message contains indication of the
```
I think @timgraham is often pushing for getting rid of starting the docstring with "Test...". Tim, can you confirm?

claudep commented:
So apart from the small remaining  question about the docstring, this looks good to me, thanks for that work!

timgraham commented on django/contrib/admin/utils.py:
```diff
@@ -490,10 +490,20 @@ def construct_change_message(form, formsets, add):
     Translation happens later on LogEntry access.
     """
     change_message = []
```
I think I would put `change_message = []` below the new code so it's not separated from where it's used.

timgraham commented on django/contrib/admin/utils.py:
```diff
@@ -490,10 +490,20 @@ def construct_change_message(form, formsets, add):
     Translation happens later on LogEntry access.
     """
     change_message = []
+    changed_data = form.changed_data
```
This alias doesn't seem needed as it's only used once.

timgraham commented on django/contrib/admin/utils.py:
```diff
@@ -490,10 +490,20 @@ def construct_change_message(form, formsets, add):
     change_message = []
+    changed_data = form.changed_data
+    changed_field_labels = []
+    with translation_override(None):
```
IMO, this block merits a comment .... "Deactivate translations  and (do something)."

timgraham commented on tests/admin_utils/admin.py:
```diff
@@ -1,11 +1,29 @@
+    def changed_data(self):
+        data = super().changed_data
+        if data:
+            # Add arbitrary name to changed_data to test change message construction.
```
Generally I like to wrap comments at 79 characters.

timgraham commented on tests/admin_views/tests.py:
```diff
@@ -6069,3 +6069,39 @@ def test_implicitly_generated_pk(self):
+
+
+@override_settings(ROOT_URLCONF='admin_views.urls')
+class AdminHistoryViewTest(TestCase):
```
Test -> Tests (and I suggest a new `test_history_view.py` file since this file is quite large already.

claudep replied:
I think reorganization of the admin views tests deserves its own patch outsie of this ticket.

CuriousLearner replied:
Shall I refactor this, or make a separate ticket for this?

timgraham replied:
Adding new code in a good location is fine.

timgraham commented on tests/admin_views/tests.py:
```diff
@@ -6069,3 +6069,39 @@ def test_implicitly_generated_pk(self):
+        changed fields as their form label.
+        """
+        state = State.objects.create(name='My State Name')
+        city = City.objects.create(name='My City Name', state_id=state.pk)
```
`state=state` is more typical than assigning `state_id`.

timgraham commented on tests/admin_views/admin.py:
```diff
@@ -881,8 +881,26 @@ class CityInlineAdmin(admin.TabularInline):
+    class Meta:
+        model = State
+        fields = '__all__'
+        labels = {'name': 'The State Name'}
```
For cases like this, I like to use semi-meaningful names that assist in debugging and understanding of how the test works, e.g. `'State name (from form's Meta.labels)'` (you could do a similar thing for the `verbose_name` additions).

CuriousLearner replied:
Yeah, that actually sounds pretty good. It would be easy to find out where it is actually coming from.

CuriousLearner commented:
Hi @timgraham 

I've addressed most of your review. There is just one small bit left out.

Here: https://github.com/django/django/pull/10531#discussion_r235206130

The issue is if I use the `form.changed_data` directly in the block where translations are disabled, then I also get `Created` field which ultimately gets appended to `changed_field_labels`

It seems like when I do take an alias but try to use `form.changed_data` directly in loop everything is working perfectly.

I'm not sure how evaluating `form.changed_data` at different times (outside and inside of block that disables translation) is causing this behavior. Do you have any idea about this?

cc @carltongibson @claudep

CuriousLearner commented:
Hey @timgraham @carltongibson @claudep 

Do you have any idea of what is going on with `form.changed_data` in this context?

CuriousLearner commented:
hmm... this is indeed interesting!

@claudep I've added a comment explaining the reasoning of evaluation of `form.changed_data` before deactivating translations.

pope1ni commented on django/contrib/admin/utils.py:
```diff
@@ -489,12 +489,32 @@ def construct_change_message(form, formsets, add):
+    # necessary to avoid conflicts in the interpretation of `datetime`
+    # fields on the models.  For example, 12/3/2018 is 12th of March for
+    # a French person, but 3rd of December for a US person (default
+    # when translations are deactivated).
```
Perhaps a less repetitive, less wordy version of this comment:
```python
    # Evaluating `form.changed_data` prior to disabling translations is required
    # to avoid fields affected by localization from being included incorrectly,
    # e.g. where date formats differ such as MM/DD/YYYY vs DD/MM/YYYY.
```

pope1ni commented on django/contrib/admin/utils.py:
```diff
@@ -489,12 +489,32 @@ def construct_change_message(form, formsets, add):
+    # a French person, but 3rd of December for a US person (default
+    # when translations are deactivated).
+    changed_data = form.changed_data
+    changed_field_labels = []
```
I believe this line can live within the with statement below without issue.

pope1ni commented on django/contrib/admin/utils.py:
```diff
@@ -489,12 +489,32 @@ def construct_change_message(form, formsets, add):
+    changed_field_labels = []
+    with translation_override(None):
+        """
+        Deactivate translations while fetching verbose_name for form
```
Should this be a non-docstring comment? (Not sure what the convention is here.)

pope1ni commented on django/contrib/admin/utils.py:
```diff
@@ -506,11 +526,19 @@ def construct_change_message(form, formsets, add):
                         }
                     })
                 for changed_object, changed_fields in formset.changed_objects:
+                    changed_fields_labels = []
```
This added block of code is essentially the same as above. Can you pull it out into a separate function that takes the form and changed fields to avoid the duplication?

pope1ni commented on docs/releases/2.2.txt:
```diff
@@ -282,6 +282,9 @@ Database backend API
 Miscellaneous
 -------------
 
+* Admin's model history change messages will prefer usage of human-readable
```
Should this come under a `django.contrib.admin` heading above? (Again, not sure on convention here.)

CuriousLearner replied:
I'm not sure about this either. @timgraham what are your thoughts on this?

CuriousLearner replied:
Thanks! I changed the message.

I believe this could be a potential breaking change if someone would be relying on iterating the dictionary of change messages of form for something.

One use-case I can think of is that being used in a test, but I believe `change_messages` can potentially have been used in code to trigger actions.

What do you think?

CuriousLearner replied:
I've added it currently in 3.0.txt breaking change section, but if you believe it's not a breaking change, then please let me know.

Thanks!

pope1ni commented on tests/admin_views/test_history_view.py:
```diff
@@ -0,0 +1,51 @@
+            'city_set-INITIAL_FORMS': '1',
+            'city_set-MAX_NUM_FORMS': '0',
+        }
+        state_change_url = reverse(
```
Will this statement will fit on a single line? (119 characters is permitted.)

pope1ni commented on tests/admin_views/test_history_view.py:
```diff
@@ -0,0 +1,51 @@
+            'admin:admin_views_state_change', args=(state.pk,)
+        )
+        self.client.post(state_change_url, change_dict)
+        logentry = LogEntry.objects.filter(
```
Will this statement will fit on a single line? (119 characters is permitted.)

pope1ni commented on tests/admin_views/test_history_view.py:
```diff
@@ -0,0 +1,51 @@
+            'nolabel_form_field and not_a_form_field. '
+            'Changed City verbose_name for city "%s".' % (
+                str(city)
+            )
```
Extra wrapping and `str()` call are unnecessary:
`… for city "%s".' % city`

CuriousLearner commented:
@pope1ni I wasn't sure about your release note entry suggestion, but I've still done it.

Do you mind having a second look here?

@timgraham Can you please have a look at this?

CuriousLearner commented on docs/releases/2.2.txt:
```diff
@@ -47,6 +47,9 @@ Minor features
 * Added a CSS class to the column headers of
   :class:`~django.contrib.admin.TabularInline`.
 
+* Admin's model history change messages will prefer usage of human-readable
```
oh, okay. Sorry I didn't get your point at first. Yeah, this makes sense.

CuriousLearner commented:
@pope1ni What do you think about this patch now?

jpic commented on django/contrib/admin/models.py:
```diff
@@ -113,6 +113,9 @@ def get_change_message(self):
+                    ]
                     sub_message['changed']['fields'] = get_text_list(
                         sub_message['changed']['fields'], gettext('and')
                     )
```
Isn't that going to inhibit the previous lines you have added ? Setting sub_message['changed']['fields'] over again twice in a row. 

Thanks for sharing some of your insight.

pope1ni replied:
Nope. In the second case `sub_message['changed']['fields']` is fetched again. This could certainty be made less clunky though as `sub_message['changed']['fields']` is referenced 4 times in 6 lines...

jpic commented on django/contrib/admin/utils.py:
```diff
@@ -489,12 +489,21 @@ def construct_change_message(form, formsets, add):
+        # Deactivate translations while fetching verbose_name for form
+        # field labels and using `field_name`, if verbose_name is not provided.
+        # Translations would happen later on LogEntry access.
+        changed_field_labels = _get_changed_field_labels_from_form(form, changed_data)
```
Use form.changed_data directly and remove the alias as recommended by Tim.

pope1ni replied:
No, this was already discussed that `form.changed_data` needed to be built prior to disabling the translations. Read the comments added above. Tim's comment is outdated but the line it is on hasn't changed so GitHub hasn't hidden it.

jpic commented on django/contrib/admin/utils.py:
```diff
@@ -521,3 +530,14 @@ def construct_change_message(form, formsets, add):
+        except KeyError:
+            verbose_field_name = field_name
+        changed_field_labels.append(str(verbose_field_name))
+    return changed_field_labels
```
I would have liked to see that as part of the public API, unless you have an objection ?

timgraham replied:
I'm not sure what you have in mind, but `django/contrib/admin/utils.py` doesn't contain any public APIs (and this code doesn't seem to be admin-specific). I think to make it a public API (e.g. in django.forms) we would want to see it used outside the admin.

carltongibson commented on django/contrib/admin/utils.py:
```diff
@@ -489,12 +489,21 @@ def construct_change_message(form, formsets, add):
+    with translation_override(None):
+        # Deactivate translations while fetching verbose_name for form
+        # field labels and using `field_name`, if verbose_name is not provided.
+        # Translations would happen later on LogEntry access.
```
"Translations **will** happen..."?

CuriousLearner commented on tests/admin_views/test_history_view.py:
```diff
@@ -0,0 +1,45 @@
+    def setUp(self):
+        self.client.force_login(self.superuser)
+
+    def test_history_view_changed_fields_message(self):
```
Yeah, that makes sense. Since we already made a new file, the name of this test should change.

CuriousLearner commented:
Hey @carltongibson 

Apologies for getting this late, but I've fixed this now.

Please let me know if there are any more changes needed.

Thanks!

CuriousLearner commented:
Hey @carltongibson 

This PR was approved but is pending for a while to be checked in.

Can you please take a look?

carltongibson commented:
Hi @CuriousLearner. Yes, super. Let's get this in 🙂  Can you rebase (rather than merge in `master`)? 

I'll have a look this week.

CuriousLearner commented:
Sorry for not being able to come to rebase this earlier and thank you for merging this in :) 🎉
